clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3783,6 +3634,327 @@ static void handleSYCLIntelMaxWorkGroupSize(Sema &S, Decl *D,
+  //
+  // The arguments to reqd_work_group_size are ordered based on which index
+  // increments the fastest. In OpenCL, the first argument is the index that
+  // increments the fastest, and in SYCL, the last argument is the index that
+  // increments the fastest.
+  //
+  // __attribute__((reqd_work_group_size)) follows the OpenCL rules in OpenCL
+  // mode. All spellings of reqd_work_group_size attribute (regardless of
+  // syntax used) follow the SYCL rules when in SYCL mode.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+    if (checkWorkGroupSizeAttrValues<std::greater<int>>(
+            *this, XDim, YDim, ZDim, DeclAttr->getXDim(), DeclAttr->getYDim(),
+            DeclAttr->getZDim())) {
+      Diag(CI.getLoc(), diag::err_conflicting_sycl_function_attributes)
+          << CI << DeclAttr;
+      Diag(DeclAttr->getLoc(), diag::note_conflicting_attribute);
+      return;
+    }
+  }
+
+  // If the 'reqd_work_group_size' attribute is specified on a declaration
+  // along with 'num_simd_work_items' attribute, the required work group size
+  // specified by 'num_simd_work_items' attribute must evenly divide the index
+  // that increments fastest in the 'reqd_work_group_size' attribute.
+  //
+  // The arguments to reqd_work_group_size are ordered based on which index
+  // increments the fastest. In OpenCL, the first argument is the index that
+  // increments the fastest, and in SYCL, the last argument is the index that
+  // increments the fastest.
+  //
+  // __attribute__((reqd_work_group_size)) follows the OpenCL rules in OpenCL
+  // mode. All spellings of reqd_work_group_size attribute
+  // (regardless of syntax used) follow the SYCL rules when in SYCL mode.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelNumSimdWorkItemsAttr>()) {
+    if (CheckWorkGroupSize(*this, DeclAttr->getValue(), XDim, ZDim)) {
+      Diag(DeclAttr->getLoc(), diag::err_sycl_num_kernel_wrong_reqd_wg_size)
+          << DeclAttr << CI;
+      Diag(CI.getLoc(), diag::note_conflicting_attribute);
+      return;
+    }
+  }
+
+  // If the attribute was already applied with different arguments, then
+  // diagnose the second attribute as a duplicate and don't add it.
+  if (const auto *Existing = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+    DupArgResult Results[] = {AreArgValuesIdentical(XDim, Existing->getXDim()),
+                              AreArgValuesIdentical(YDim, Existing->getYDim()),
+                              AreArgValuesIdentical(ZDim, Existing->getZDim())};
+    // If any of the results are known to be different, we can diagnose at this
+    // point and drop the attribute.
+    if (llvm::is_contained(Results, DupArgResult::Different)) {
+      Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+      Diag(Existing->getLoc(), diag::note_previous_attribute);
+    }
+
+    // If the 'reqd_work_group_size' attribute is specified multiple times on a
```

Comment:
I don't understand this. If attribute is already applied with different arguments, won't we diagnose this above? Can you point me to the test where this is not the case?

Reply:
>>If attribute is already applied with different arguments, won't we diagnose this above?

Yes , we do diagnose.

>>Can you point me to the test where this is not the case?

No, we do not have any case for this.

Here are differences that match with existing implementation.

Example: Both diagnose will be seen (duplicate + conflict)
```
[[sycl::reqd_work_group_size(4, 4, 4)]] void func3();   // expected-note {{previous attribute is here}} \
		                                        // expected-note {{conflicting attribute is here}}
[[sycl::reqd_work_group_size(1, 1, 1)]] void func3() {} // expected-warning {{attribute 'reqd_work_group_size' is already applied with different arguments}} \
		                                        // expected-error {{'reqd_work_group_size' attribute conflicts with 'reqd_work_group_size' attribute}}
```
Only duplicate diagnose will be seen for the test case below.
```
[[sycl::reqd_work_group_size(1, 1, 1)]] void func3();   // expected-note {{previous attribute is here}}                                        
[[sycl::reqd_work_group_size(2, 2, 2)]] void func3() {} // expected-warning {{attribute 'reqd_work_group_size' is already applied with different arguments}} 
```

Reply:
Oh I see. Strange design... Is this new functionality? I don't see where this is done in old code. Could you please point me to it?

Reply:
No it is old functionality:
https://github.com/intel/llvm/pull/5782/files#diff-21cade49aeba756014c248b8c00eda9b3c0a74d6af5cb32b97875bb5ac1d4114R30

Reply:
Right. I understand we throw error for conflicting attribute arguments. My question was about the argument check - 
```
// check if the values of 'reqd_work_group_size' attribute
// arguments specified earlier are less than the values of
// 'reqd_work_group_size' attribute arguments specified after.
```
I was wondering where this was done before.

Reply:
Are you taking about different spellings of  reqd_work_group_size' attribute?

Reply:
Unless I'm missing something here, I don't think we need a warning and error for the same issue. I think you can just change the warning diagnostic above to an error and remove this code. 

We can also then rename `checkWorkGroupSizeAttrValues` to `checkMaxWorkGroupSize` and maybe remove comparator, since that function then is just used to check if attribute values are below allowed maximum.

Reply:
>> I don't think we need a warning and error for the same issue. 

I thought about this. To keep original changes, I did not change this. Yes, it makes sense to me too we do not need a warning and error for the same issue.

>>I think you can just change the warning diagnostic above to an error and remove this code.

Done

>>e can also then rename checkWorkGroupSizeAttrValues to checkMaxWorkGroupSize and maybe remove comparator, since that function then is just used to check if attribute values are below allowed maximum.

@elizabethandrews, I have decided to keep checkWorkGroupSizeAttrValues() since we are comparing maximum values between reqd_work_group_size and max_work_group_size attributes.

Reply:
I see where the `checkMaxWorkGroupSize` might be confusing since we have the an attribute of same name. I didn't think of that when I suggested name. WDYT about renaming it to `checkMaximumWorkGroupSize` or `checkMaxAllowedWorkGroupSize`, since that is essentially what function is testing. I just think its more readable to name a function based on what it does especially because we have other functions with similar names

Reply:
I have kept it like checkWorkGroupSizeAttrValues() since we are checking mutual diagnostics of different workgroup size attribute by comparing maximum values between reqd_work_group_size and max_work_group_size attributes.

Reply:
It is my understanding that this function checks the maximum allowed values for work group size by comparing that 'required' size (specified by reqd_work_group_size) does not exceed maximum allowed size (specified by `max_work_group_size`). So I suggested just calling it `checkMaxAllowedWorkGroupSize`, since that describes the functionality, and we already have other functions with similar generic names like `InvalidWorkGroupSizeAttrs` and `CheckWorkGroupSize`. IMO all these names are similar and don't really specify what we are checking. Anyway, it's not a big deal. Its just a personal preference at this point. So if you disagree with me, you don't need to change it! I won't block PR because of it.

Reply:
Makes sense! I have updated it. Thank you